core/src/main/scala/org/apache/spark/rpc/netty/Outbox.scala

```diff
@@ -66,14 +66,22 @@ private[netty] case class RpcOutboxMessage(
     this.requestId = client.sendRpc(content, this)
   }
 
-  def onTimeout(): Unit = {
+  private[netty] def removeRpcRequest(): Unit = {
     if (client != null) {
       client.removeRpcRequest(requestId)
     } else {
-      logError("Ask timeout before connecting successfully")
+      logError("Ask terminated before connecting successfully")
     }
   }
 
+  def onTimeout(): Unit = {
+    removeRpcRequest()
+  }
+
+  def onAbort(): Unit = {
+    removeRpcRequest()
```

Comment:
why don't `onFailure` call `removeRpcRequest`?

Reply:
Not a good idea.
When we want to abort the RPC, we should throw an RPC aborted exception (thrown from the `Future` thread itself). And let the `Future` thread to catch the exception, and do cleanup, and then re-throw the exception.

We can also reference how RPC impl the "timeout" , see
```
      val timeoutCancelable = timeoutScheduler.schedule(new Runnable {
        override def run(): Unit = {
          onFailure(new TimeoutException(s"Cannot receive any reply from ${remoteAddr} " +
            s"in ${timeout.duration}"))
        }
      }, timeout.duration.toNanos, TimeUnit.NANOSECONDS)
```
It only call `onFailure(new TimeoutException(...))` and inside `onFailure` it call `promise.tryFailure(e)` to raise exception `Future` thread itself.